packages/plugin-file/sources/fileUtils.ts

```diff
@@ -0,0 +1,77 @@
+import {structUtils, FetchOptions, Locator, miscUtils, tgzUtils, Ident} from '@yarnpkg/core';
+import {ppath, PortablePath, npath, CwdFS, ZipFS, xfs}                  from '@yarnpkg/fslib';
+
+export function parseSpec(spec: string) {
+  const {params, selector} = structUtils.parseRange(spec);
+
+  const path = npath.toPortablePath(selector);
+
+  const parentLocator = params && typeof params.locator === `string`
+    ? structUtils.parseLocator(params.locator)
+    : null;
+
+  return {parentLocator, path};
+}
+
+export function makeSpec({parentLocator, path, folderHash, protocol}: {parentLocator: Locator | null, path: string, folderHash?: string, protocol: string}) {
+  const parentLocatorSpread = parentLocator !== null
+    ? {locator: structUtils.stringifyLocator(parentLocator)}
+    : {} as {};
+
+  const folderHashSpread = typeof folderHash !== `undefined`
+    ? {hash: folderHash}
+    : {} as {};
+
+  return structUtils.makeRange({
+    protocol,
+    source: path,
+    selector: path,
+    params: {
+      ...folderHashSpread,
+      ...parentLocatorSpread,
+    },
+  });
+}
+
+export function makeLocator(ident: Ident, {parentLocator, path, folderHash, protocol}: Parameters<typeof makeSpec>[number]): Locator {
+  return structUtils.makeLocator(ident, makeSpec({parentLocator, path, folderHash, protocol}));
+}
+
+export async function loadFolderFs(locator: Locator, protocol: string, opts: FetchOptions): Promise<ZipFS> {
```

Comment:
I'd prefer to pass the options using a map object, and perhaps to give a more obvious name to the function (`load` doesn't convey that we're generating something). For example:
```suggestion
export async function makeArchiveFromParent(locator: Locator, opts: {protocol: string, fetchOptions: FetchOptions}): Promise<ZipFS> {
```

Reply:
Done! I went with `makeArchiveFromLocator` since it's even more descriptive.